### ISSUE TITLE
string index out of range in `replace_part_with_missing_leading_whitespace`

### ISSUE THREAD

**Issue by wwestenbrink** [paul-gauthier/aider#25]
hi Paul,

Noticed another crash when applying another java diff.

```
string index out of range

Traceback (most recent call last):
  File "/Users/wouter/Projects/try/aider/aider/coder.py", line 807, in apply_updates
    edited = method(content)
             ^^^^^^^^^^^^^^^
  File "/Users/wouter/Projects/try/aider/aider/coder.py", line 620, in update_files_gpt4
    if utils.do_replace(full_path, original, updated, self.dry_run):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/wouter/Projects/try/aider/aider/utils.py", line 216, in do_replace
    new_content = replace_most_similar_chunk(content, before_text, after_text)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/wouter/Projects/try/aider/aider/utils.py", line 99, in replace_most_similar_chunk
    res = replace_part_with_missing_leading_whitespace(whole, part, replace)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/wouter/Projects/try/aider/aider/utils.py", line 76, in replace_part_with_missing_leading_whitespace
    if c == part_lines[0][0]:
            ~~~~~~~~~~~~~^^^
```

This was the diff it generated, seems it has a problem with the empty line in the beginning

```
 src/test/java/de/wwestenbrink/demo/scheduler/SchedulerNotificationTest.java
 <<<<<<< ORIGINAL

   @Test
   public void shouldNotHaveAccessToOthersNotification() {
     String notificationId = scheduleNotification(60, "other@test.com");

     List<NotificationDTO> response = listNotifications(TEST_SUBJECT);
     assertThat(response).isEmpty();

     getNotification(notificationId, TEST_SUBJECT, HttpStatus.NOT_FOUND);

     cancelNotification(notificationId, TEST_SUBJECT, HttpStatus.NOT_FOUND);
   }
 =======
   @Test
   public void shouldNotHaveAccessToOthersNotification() {
     String notificationId = scheduleNotification(60, "other@test.com");

     List<NotificationDTO> response = listNotifications(TEST_SUBJECT);
     assertThat(response).isEmpty();

     getNotification(notificationId, TEST_SUBJECT, HttpStatus.NOT_FOUND);

     cancelNotification(notificationId, TEST_SUBJECT, HttpStatus.NOT_FOUND);
   }

   @Test
   public void shouldNotFindNonExistingNotification() {
     String nonExistingNotificationId = "nonExistingId";
     getNotification(nonExistingNotificationId, TEST_SUBJECT, HttpStatus.NOT_FOUND);
   }
 >>>>>>> UPDATED
 ```
 
Not sure about the fix this time.

I've noticed aider is struggling a bit with diffs in general, have you considered using Unified Format (unidiff)?

**Comment by paul-gauthier** [comment]
Good catch, thanks for reporting. I just pushed a fix. You'll need to install from github to pick it up.

The root of the issue is that GPT-4 won't consistently return properly formatted diffs. The <<ORIGINAL/>>UPDATED format aider uses is by far the most reliable format I've discovered. But GPT-4 won't always conform and output even this very simple format.

I've tried asking it to output `diff -c`, `diff -u`, `sed`, and many other formats. GPT-4 does much, much worse on them. Even the new functions api is not looking like a panacea here.

One key issue about the "output format" is that it's easy to "distract" the LLM if you ask it to use too structured or technical of a format. Especially when you're asking it to write code. The coding itself is a challenging task, and asking it to render actual diff or vim or sed editing outputs is going to confuse it. I have absolutely seen the quality of the coding task degrade when the output format is too complex.

Imagine asking a junior developer to "implement this cool feature, but I want you to hand type the diffs required to change the current code".

I'm hoping to write up something more quantitative and concrete about this aspect of AI coding soon.
